src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/****************************************************************************
+ * @file
+ * @brief Implementation for the Administrator Commissioning Cluster
+ ***************************************************************************/
+
+#include <app/CommandHandler.h>
+#include <app/server/Server.h>
+#include <app/util/af.h>
+#include <setup_payload/SetupPayload.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip;
+
+// Specifications section 5.4.2.3. Announcement Duration
+constexpr uint32_t kMaxCommissionioningTimeoutSeconds = 15 * 60;
+
+bool emberAfAdministratorCommissioningClusterOpenCommissioningWindowCallback(chip::EndpointId endpoint,
+                                                                             app::CommandHandler * commandObj,
+                                                                             uint16_t commissioningTimeout, ByteSpan pakeVerifier,
+                                                                             uint16_t discriminator, uint32_t iterations,
+                                                                             ByteSpan salt, uint16_t passcodeID)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    PASEVerifier verifier;
+    const uint8_t * verifierData = pakeVerifier.data();
+
+    ChipLogProgress(Zcl, "Received command to open commissioning window");
+
+    VerifyOrExit(!IsPairingWindowOpen(), status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(sizeof(verifier) == pakeVerifier.size(), status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(iterations >= kPBKDFMinimumIterations, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(iterations <= kPBKDFMaximumIterations, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(salt.size() >= kPBKDFMinimumSaltLen, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(salt.size() <= kPBKDFMaximumSaltLen, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(commissioningTimeout <= kMaxCommissionioningTimeoutSeconds, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(discriminator <= kMaxDiscriminatorValue, status = EMBER_ZCL_STATUS_FAILURE);
+
+    memcpy(&verifier[0][0], &verifierData[0], kSpake2p_WS_Length);
```

Comment:
Any chance we could make PASEVerifier not be a 2D array and instead name it in some way? 

What is verifier[0] and what is verifier[1]? it would be easier to say "verifier.friendlyNameHere" instead of verifier[0].

&verifier[0][0] seems to be just `verifier[0]`. Why the extra indirections?

Reply:
I'll try to roll it as part of this PR.
`&verifier[0][0]` was a try to make it more readable. Once we change it to friendly named structure, it'll get ironed out.